pilot/proxy/envoy/config.go

```diff
@@ -934,13 +939,98 @@ func buildEgressHTTPRoutes(mesh *meshconfig.MeshConfig, node proxy.Node,
 				Port: intPort, Protocol: protocol}
 			httpConfig := httpConfigs.EnsurePort(intPort)
 			httpConfig.VirtualHosts = append(httpConfig.VirtualHosts,
-				buildEgressVirtualHost(rule, mesh, node, modelPort, instances, config))
+				buildEgressVirtualHost(rule.Destination.Service, mesh, node, modelPort, modelPort, instances, config))
 		}
 	}
 
 	return httpConfigs.normalize()
 }
 
+
+func buildForeignServiceHTTPRoutes(mesh *meshconfig.MeshConfig, node proxy.Node,
+	instances []*model.ServiceInstance, config model.IstioConfigStore,
+	httpConfigs HTTPRouteConfigs) HTTPRouteConfigs {
+
+	if node.Type == proxy.Router {
+		// No egress rule support for Routers. As semantics are not clear.
+		return httpConfigs
+	}
+
+	fsConfigs := config.ForeignServices()
+	for _, fsConfig := range fsConfigs {
+		fs := fsConfig.Spec.(*routingv2.ForeignService)
+		for _, port := range fs.Ports {
+			protocol := model.ConvertCaseInsensitiveStringToProtocol(port.Protocol)
+			if !model.IsEgressRulesSupportedHTTPProtocol(protocol) {
+				continue
+			}
+
+			// TODO: in/out port logic needs to be tested
+			outPortInt := int(port.OutPort)
+			if outPortInt == 0 {
+				outPortInt = int(port.InPort)
+			}
+
+			inPort := &model.Port{Name: fmt.Sprintf("external-%v-%d", protocol, int(port.InPort)),
+				Port: int(port.InPort), Protocol: protocol}
+			outPort := &model.Port{Name: fmt.Sprintf("external-%v-%d", protocol, outPortInt),
+				Port: outPortInt, Protocol: protocol}
+
+			httpConfig := httpConfigs.EnsurePort(int(port.InPort))
+			for _, host := range fs.Hosts {
+				httpConfig.VirtualHosts = append(httpConfig.VirtualHosts,
+					buildEgressVirtualHost(host, mesh, node, inPort, outPort, instances, config))
+			}
+		}
+	}
+
+	return httpConfigs.normalize() // TODO: is this necessary?
+}
+
+func buildForeignServiceTCPListeners(mesh *meshconfig.MeshConfig, node proxy.Node,
+	config model.IstioConfigStore) (Listeners, Clusters) {
+
+	listeners := make(Listeners, 0)
+	clusters := make(Clusters, 0)
+
+	if node.Type == proxy.Router {
+		return listeners, clusters
+	}
+
+	// TODO: port conflicts?
+	for _, fsConfig := range config.ForeignServices() {
+		fs := fsConfig.Spec.(*routingv2.ForeignService)
+		for _, port := range fs.Ports {
+			protocol := model.ConvertCaseInsensitiveStringToProtocol(port.Protocol)
+			if !model.IsEgressRulesSupportedTCPProtocol(protocol) {
```

Comment:
why? We can still do plain tcp proxy right? In fact, thats all we would do, except for mongo

Reply:
this filters out http, http2 and grpc ports. yes, we only support tcp and mongo here.